Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,223 @@
+        }
+
+        /// Get an option argument's value from the results.
+        ///
+        /// Since the options are optional, their result may or may not
+        /// be present.
+        public func get<T>(_ arg: OptionArgument) -> T? {
+            return results[arg] as? T
+        }
+
+        /// Get a positional argument's value.
+        public func get<T>(_ arg: PositionalArgument) -> T {
+            return results[arg] as! T
+        }
+    }
+
+    /// List of arguments added to this parser.
+    private var arguments = [CommandLineArgument]()
+
+    public init() {
+    }
+
+    /// Adds an option to the parser.
+    public mutating func add(option: String, kind: CommandLineArgument.Kind, usage: String? = nil) -> OptionArgument {
+        let arg = OptionArgument(name: option, kind: kind, usage: usage)
+        arguments.append(arg)
+        return arg
+    }
+
+    /// Adds an argument to the parser.
+    public mutating func add(argument: String, kind: CommandLineArgument.Kind, usage: String? = nil) -> PositionalArgument {
+        let arg = PositionalArgument(name: argument, kind: kind, usage: usage)
+        arguments.append(arg)
+        return arg
+    }
+
+    /// Parses the provided array and return the result.
+    public mutating func parse(_ args: [String] = []) throws -> ArgumentParserResult {
+        var result = ArgumentParserResult()
+
+        // Create options map to quickly look up the arguments.
+        let optionsMap = Dictionary(items: arguments.flatMap { $0 as? OptionArgument }.map { ($0.name, $0) })
+        // Seperate out positional arguments.
+        let positionalArgs = arguments.flatMap { $0 as? PositionalArgument }
+        // Create iterators.
+        var positionalArgsIterator = positionalArgs.makeIterator()
+        var it = args.makeIterator()
+
+        while let arg = it.next() {
+            if arg.isPositionalArg {
+                guard let positionalArg = positionalArgsIterator.next() else {
+                    throw ArgumentParserError.unexpectedArgument(arg)
+                }
+                try result.addResult(for: positionalArg, value: arg)
+            } else {
+                guard let option = optionsMap[arg] else {
+                    throw ArgumentParserError.unknown(option: arg)
+                }
+                // FIXME: Find a better way to handle bool which doesn't have associated values.
+                if option.kind == .bool {
+                    try result.addResult(for: option, value: "")
+                } else if let value = it.next() {
+                    try result.addResult(for: option, value: value)
+                } else {
+                    throw ArgumentParserError.expectedValue(option: arg)
+                }
+            }
+        }
+        // Report if there are any positional arguments which were not present in the arguments.
+        if let leftOverPositionalArgument = positionalArgsIterator.next() {
```

Comment:
Could just do `leftOverArgs = [String](positionalArgsIterator)` and then check if empty.